TE-14.2 gribi encap and decap scale

cprabha opened the pull request:
Hi,

This is initial commit for TE-14.2.

Resolves: https://github.com/openconfig/featureprofiles/issues/2223

Needs https://github.com/openconfig/featureprofiles/pull/2702/files to be merged.

Thanks,
Prabha

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
````diff
@@ -0,0 +1,312 @@
+    * proto: `41`    
+    ```
+
+3.  Send traffic to DUT-1, covering all the installed v4 and v6 entries in the decap and encap VRFs. Validate that all traffic are all decapped per the DECAP VRFs and then encapsulated per the ENCAP VRFs and received as encapsulated packet by ATE.
````
I _think_ this means that the `ipv*_inner_src` and `ipv*_inner_encap_match` variables have to come from these prefixes, but it would be helpful if the text was more explicit about this.

cprabha replied:
This README is given by google.

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
````diff
@@ -0,0 +1,312 @@
+    ```
+    * inner_src: `ipv4_inner_src`
+    * inner_dst: `ipv4_inner_encap_match`
+    * dscp: `dscp_encap_a`
````
For all of these packet templates, should this line be `inner_dscp` and the next one `outer_dscp`, to be more clear?

cprabha replied:
If it below Inner_src/Dst means inner header dscp bit. ?

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+                            sequence-id: 1
+                            ipv4 {
+                                protocol: 4
+                                dscp-set: [dscp_encap_a_1, dscp_encap_a_2]
```
(obviously, make sure that these DSCP values are the ones that you want to use... I just used them as an example)

cprabha replied:
update , plz check

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
````diff
@@ -0,0 +1,312 @@
+    ```
+
+3.  Send traffic to DUT-1, covering all the installed v4 and v6 entries in the decap and encap VRFs. Validate that all traffic are all decapped per the DECAP VRFs and then encapsulated per the ENCAP VRFs and received as encapsulated packet by ATE.
+4.  Flush the `DECAP_TE_VRF`, install 5000 entries with fixed prefix length of /32, and repeat the same traffic validation.
````
What are the values (prefix or range) to be used for these 5000 /32 entries? How are they related to the other prefixes used in this test, e.g. are they meant to cover the same range as the 48 prefixes used for `DECAP_TE_VRF` in step 1 above?

cprabha replied:
I have provided these variables in begining of the test,
IPBlockEncapA           = "20.1.64.1/15"  // IPBlockEncapA represents the ipv4 entries in EncapVRFA
	IPBlockEncapB           = "20.5.64.1/15"  // IPBlockEncapB represents the ipv4 entries in EncapVRFB
	IPBlockEncapC           = "20.10.64.1/15" // IPBlockEncapC represents the ipv4 entries in EncapVRFC
	IPBlockEncapD           = "20.15.64.1/15" // IPBlockEncapD represents the ipv4 entries in EncapVRFD
	IPBlockDecap            = "30.0.0.1/15"   // IPBlockDecap represents the ipv4 entries in Decap VRF
	ipv4OuterSrc111         = "198.51.100.111"
	gribiIPv4EntryVRF1111   = "203.0.113.1"
	IPv6BlockEncapA         = "2001:DB8:0:1::/64"
	IPv6BlockEncapB         = "2001:DB8:1:1::/64"
	IPv6BlockEncapC         = "2001:DB8:2:1::/64"
	IPv6BlockEncapD         = "2001:DB8:3:1::/64"

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -0,0 +1,1397 @@
+	IPBlockDefaultVRF       = "198.18.128.0/18"
+	IPBlockNonDefaultVRF    = "198.18.0.0/17"
+	tunnelSrcIP             = "198.51.100.99" // tunnelSrcIP represents Source IP of IPinIP Tunnel
+	policyName              = "redirect-to-VRF1"
```
Is this policy used in this test?

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -0,0 +1,1397 @@
+//   - ate:port2.i -> dut:port2.i VLAN-ID i subnet 198.18.x.(4*i)/30 (out of subnet 198.18.192.0/18)
+const (
+	ipv4PrefixLen           = 30 // ipv4PrefixLen is the ATE and DUT interface IP prefix length
+	vrf1                    = "VRF-A"
```
I see this variable used in code below, to push routes. But it's not described in the README, and I'm unsure if it's a typo in the code?

cprabha replied:
As per requirement we are building 14.2 decap/encap scale on top of 14.1. Hence the ask was to retain 14.1 scale.

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -0,0 +1,1397 @@
+//   - ate:port2.1 -> dut:port2.1 VLAN-ID: 1 subnet 198.18.192.4/30
+//   - ate:port2.2 -> dut:port2.2 VLAN-ID: 2 subnet 198.18.192.8/30
+//   - ate:port2.i -> dut:port2.i VLAN-ID i subnet 198.18.x.(4*i)/30 (out of subnet 198.18.192.0/18)
+const (
```
As mentioned in a comment on the README, I suggest mentioning these variables in the README text if they're also being used in templates in the README. The example I gave was for DSCP values, but it might apply to other variables here?

cprabha replied:
This README is provided by google, I have updated for VRF selection policy. We usually updated in README briefly for some variables not all of variables which are used in scripts.

bensons commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -0,0 +1,1397 @@
+	niTeVrf111              = "TE_VRF_111"
+	niTeVrf222              = "TE_VRF_222"
+	niDefault               = "DEFAULT"
+	IPBlockEncapA           = "20.1.64.1/15"  // IPBlockEncapA represents the ipv4 entries in EncapVRFA
```
Should we use different IP address prefixes, such as private addresses (RFC 1918) or the TEST-NET blocks reserved in RFC 5735? The prefixes defined in this code are registered to existing users on the Internet (e.g. Microsoft) and it would be safer to use ranges that would be captured by bogon filters, just in case...

cprabha replied:
AS Per contributing.md, we only have three ip subnets.  Out of these we already have used 2 subnets. One for interface ip addressing and other one for VRF-A/B/C.
Remaining address space is /24 which does not provide addresses which are needed for scaling.
Hence we are using other ip blocks.

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+## Procedure
+
+1.  via gRIBI installs the following AFT entries:
+    *   Add 4 VRFs for encapsulations: `ENCAP_TE_VRF_A`, `ENCAP_TE_VRF_B`, `ENCAP_TE_VRF_C` and `ENCAP_TE_VRF_D`.
```
updated, please check

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+                            ipv4 {
+                                protocol: 4
+                                dscp-set: [dscp_encap_a_1, dscp_encap_a_2]
+                                source-address: "ipv4_outer_src_222"
```
updated, please chec,

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+
+## Procedure
+
+1.  via gRIBI installs the following AFT entries:
```
Yes ,  We need to configure network instances and then apply policy.

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+
+1.  via gRIBI installs the following AFT entries:
+    *   Add 4 VRFs for encapsulations: `ENCAP_TE_VRF_A`, `ENCAP_TE_VRF_B`, `ENCAP_TE_VRF_C` and `ENCAP_TE_VRF_D`.
+    *   Inject 5000 IPv4Entry-ies and 5000 IPv6Entry-ies to each of the 4 encap VRFs.
```
14.1 is scale script and 14.2 is decap/encap scale script.
14.2 README is provided by google , as per readme provided are using 14.1 and adding 14.2 steps on top of it.

We are installing unique prefixes in each encap VRF pointing to TE_VRF111 tunnels, TE_VRF_111 utilizes base ipv4 entries added as in 14.1

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+    * inner_src: `ipv4_inner_src`
+    * inner_dst: `ipv4_inner_encap_match`
+    * dscp: `dscp_encap_a`
+    * outter_src: `ipv4_outter_src_222`
```
updated, please check.

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/README.md:
```diff
@@ -0,0 +1,312 @@
+    *   The entries in the encap VRFs should point to NextHopGroups in the `DEFAULT` VRF. Inject 200 such NextHopGroups in the DEFAULT VRF.
+    *   Each NextHopGroup should have 8 NextHops where each NextHop points to a tunnel in the `TE_VRF_111`. In addition, the weights specified in the NextHopGroup should be co-prime and the sum of the weights should be 16.
+    *   Add 1 VRF for decapsulation, `DECAP_TE_VRF`.
+    *   Inject `48` entries in the DECAP_TE_VRF where the entries have a mix of prefix lengths /22, /24, /26, and /28.
```
48 entries are for DECAP VRF.  5k Entries in Encap VRF. These are used based on VRF Selection selection policy for packet forwarding.

mohanasm commented:
Resolves https://github.com/openconfig/featureprofiles/issues/2223

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -561,27 +561,33 @@ func createAndSendTrafficFlows(t *testing.T, args *testArgs, decapEntries []stri
+	// Below code will be un commented once ixia issue is fixed.
+	// https://github.com/open-traffic-generator/fp-testbed-juniper/issues/49
+	/*
+		flow5 := createFlow(&flowArgs{flowName: "flow5", isInnHdrV4: false, outHdrDstIPCount: decapRouteCount,
```
@xw-g , I have un-commented code .  I think to access ixia issue you need to access from ixia for that repo.

Here is the  description I have added in ixia issue.

Hi,

I am sending ipv6 flow with traffic class 18, 28, 38 and 48 from OTG. But they are received on DUT as 4, 7,9 and 12.
is there any conversion happening internally inside otg? We want to receive the values as we set in script otherwise firewall
match wont happen.

Code used in script:

if len(flowValues.inHdrDscp) != 0 {
innerIpv6Hdr.TrafficClass().SetValues(flowValues.inHdrDscp)
}

cprabha commented on feature/experimental/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -599,7 +605,11 @@ func createAndSendTrafficFlows(t *testing.T, args *testArgs, decapEntries []stri
+
+	flowNameList := []string{"flow1", "flow2", "flow3", "flow4"}
 
 	verifyTraffic(t, args, flowNameList, !wantLoss)
```
We only checking for pass traffic in this script, hence removed this part of code .

xw-g commented:
@cprabha can you help fix the pull check failing? (using IP outside the ranges) Thanks a lot.

cprabha commented:
> @cprabha can you help fix the pull check failing? (using IP outside the ranges) Thanks a lot.

@xw-g Please could you merge https://github.com/openconfig/featureprofiles/pull/2702 ?
It has ip range specified from 20.0, and changed this to match 20.x, . this should fix the issue.

cprabha commented:
> > > @cprabha can you help fix the pull check failing? (using IP outside the ranges) Thanks a lot.
> > 
> > 
> > @xw-g Please could you merge #2702 ? It has ip range specified from 20.0, and changed this to match 20.x, . this should fix the issue.
> 
> It's not just a check-script issue. (you are right the the script does have issue, it's checking for `20.0/16`) The contribution guide is about `20.0/15`. You are using `20.5` and `20.10`, which is outside the `20.0/15` range.

Updated IP address range to Martian addresses ,  Please check.
